src/crimson/os/seastore/seastore_types.h

```diff
@@ -581,6 +753,92 @@ struct seg_paddr_t : public paddr_t {
       return *this;
   }
 };
+
+struct blk_paddr_t : public paddr_t {
+  static constexpr uint64_t BLK_OFF_MASK = std::numeric_limits<uint32_t>::max() 
+    >> (std::numeric_limits<uint32_t>::digits - BLK_OFF_LEN_BITS);
+  // mask for random block manager id
+  static constexpr uint64_t BLK_ID_MASK =
+    static_cast<common_addr_t>(0xFFFFFFFFFFF) << BLK_OFF_LEN_BITS;
+
+  blk_paddr_t(const blk_paddr_t&) = delete;
+  blk_paddr_t(blk_paddr_t&) = delete;
+  blk_paddr_t& operator=(const blk_paddr_t&) = delete;
+  blk_paddr_t& operator=(blk_paddr_t&) = delete;
+  block_id_t get_block_id() const {
+    return block_id_t((dev_addr & BLK_ID_MASK) >> BLK_OFF_LEN_BITS);
+  }
+  block_off_t get_block_off() const {
+    return block_off_t(dev_addr & BLK_OFF_MASK);
+  }
+  void set_block_id(const block_id_t id) {
+    dev_addr &= static_cast<common_addr_t>(
+      std::numeric_limits<device_block_id_t>::max());
+    dev_addr |= static_cast<common_addr_t>(id.block) << BLK_OFF_LEN_BITS;
+  }
+  void set_block_off(const block_off_t off) {
+    dev_addr &= static_cast<common_addr_t>(
+      std::numeric_limits<device_block_id_t>::max()) << BLK_OFF_LEN_BITS;
+    dev_addr |= ((uint32_t)off.get_raw_bits() >> 12);
+  }
+
+  paddr_t add_offset(block_off_t o) const {
+    return paddr_t::make_blk_paddr(get_block_id(), get_block_off() + o);
+  }
+
+  paddr_t add_relative(paddr_t o) const {
+    assert(o.is_relative());
+    blk_paddr_t& s = o.as_blk_paddr();
```

Comment:
I don't think we want separate representations for relative addresses between block and segment -- they are semantically identical.